tun/tun_solaris.go

```diff
@@ -0,0 +1,398 @@
+		}
+		statusUp = up
+
+		// MTU changes
+		if iface.MTU != statusMTU {
+			tun.events <- EventMTUUpdate
+		}
+		statusMTU = iface.MTU
+	}
+}
+
+func tunDestroy(name string) error {
+	var l unix.Lifreq
+	if err := l.SetName(name); err != nil {
+		return err
+	}
+	// cleanup both ipv4 and ipv6
+	for _, proto := range []int{unix.AF_INET, unix.AF_INET6} {
+		fd, err := unix.Socket(proto, unix.SOCK_DGRAM, 0)
+		if err != nil {
+			return fmt.Errorf("could not open Socket of type %v", proto)
+		}
+		defer unix.Close(fd)
+
+		// get the STREAMS muxid for the tun device stream
+		reqnum := int(unix.SIOCGLIFMUXID)
+
+		// solaris/illumos defines the ioctl number as a signed int, but the
+		// common x/sys/unix functions use uint.
+		if err := unix.IoctlLifreq(fd, uint(reqnum), &l); err != nil {
+			continue
+		}
+
+		id := l.GetLifruInt()
+
+		// Unlink the tun stream from the IP mutiplexor
+		if err := unix.IoctlSetInt(fd, unix.I_PUNLINK, id); err != nil {
+			continue
+		}
+	}
+	return nil
+}
+
+func CreateTUN(name string, mtu int) (Device, error) {
+	ppa := -1
+	if name != "tun" {
+		_, err := fmt.Sscanf(name, "tun%d", &ppa)
+		if err != nil || ppa < 0 {
+			return nil, fmt.Errorf("interface name must be tun[0-9]*")
+		}
+	}
+
+	fd, err := unix.Open(tunNode, unix.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("could not open TUN (%s)", tunNode)
+	}
+
+	// The "tun" device is a STREAMS module.  We need to make a STREAMS ioctl
+	// request to that module using the TUNNEWPPA command.  Returns the newly
+	// allocated PPA number. Passing -1 will request a dynamically assinged
+	// PPA. Passing a specific id will try to use the one specified.
+	var s unix.Strioctl
+	s.Cmd = unix.TUNNEWPPA
+	s.SetInt(ppa)
+	ppa, err = unix.IoctlSetStrioctlRetInt(fd, unix.I_STR, &s)
+	if err != nil {
+		return nil, err
+	}
+
+	assignedName := fmt.Sprintf("tun%d", ppa)
+
+	// Clean up anything possibly left behind by a crash
+	tunDestroy(assignedName)
+
+	// The solaris/illumos tun driver doesn't have an ioctl for reading back the PPA.
+	// Just pass it through to the daemon process via environment variable.
+	err = os.Setenv("_WG_TUN_NAME", assignedName)
+	if err != nil {
+		return nil, err
+	}
+	file := os.NewFile(uintptr(fd), "")
+
+	tun, err := CreateTUNFromFile(file, mtu)
+
+	if err == nil && name == "tun" {
+		fname := os.Getenv("WG_TUN_NAME_FILE")
+		if fname != "" {
+			os.WriteFile(fname, []byte(tun.(*NativeTun).name+"\n"), 0o400)
+		}
+	}
+
+	return tun, err
+}
+
+func CreateTUNFromFile(file *os.File, mtu int) (Device, error) {
+	// The solaris/illumos tun driver doesn't have an ioctl for reading back the PPA.
+	// Just get it back via environment variable.
+	name := os.Getenv("_WG_TUN_NAME")
+	if name == "" {
+		return nil, fmt.Errorf("could not determine device name")
+	}
+
+	// By holding open this file descriptor on /dev/ip for the life of the connection
+	// and using I_LINK, STREAMS will tear everything down cleanly when we close these
```

Comment:
But then I_LINK isn't actually set until deep down in a separate function. Maybe configureSTREAMS should be inlined into this function instead of being standalone, so that it's more clear.